lib/timeline/timeline.go

```diff
@@ -0,0 +1,280 @@
+// change in the cluster status. An event could be triggered by a failed
+// heartbeat or a failed health check.
+type Event struct {
+	// TimeStamp specifies when the event occurred.
+	TimeStamp time.Time
+
+	// TODO: can we rely on user CLI sessions to provide colored text?
+	// If not, maybe we should indicate the different event types a different
+	// way.
+
+	// Color specifies the color of the event.
+	// Red -> Indicates change has caused state to degraded.
+	// Yellow -> Indicates change has not caused any state changes.
+	// Green -> Indicates changes has caused state to recover.
+	Color string
+
+	// Description specifies a description of the event.
+	Description string
+}
+
+// NewEvent initializes and returns a new Event with the current timestamp.
+func NewEvent() Event {
+	return Event{
+		TimeStamp: time.Now(),
+	}
+}
+
+// String returns a string representation of Event.
+func (e *Event) String() string {
+	var sb strings.Builder
+	sb.WriteString(fmt.Sprintf("[%s] ", e.TimeStamp.Format(Stamp)))
+	sb.WriteString(e.Description)
+	return sb.String()
+}
+
+// Timeline represents a timeline of cluster status events. The Timeline
+// can hold a specified amount of events and uses a FIFO eviction policy.
+type Timeline struct {
+	// Size specifies the max size of the timeline.
+	Size int
+	// Timeline holds the latest status events.
+	Timeline []string
+	// Cluster holds the latest cluster status.
+	Cluster *Cluster
+}
+
+// NewTimeline initializes and returns a new StatusTimeline with the
+// specified size.
+func NewTimeline(size int) Timeline {
+	return Timeline{
+		Size:     size,
+		Timeline: []string{},
+		Cluster:  &Cluster{},
+	}
+}
+
+// RecordStatus records differences of the previous status to the provided
+// status into the Timeline.
+func (t *Timeline) RecordStatus(status *pb.SystemStatus) {
+	cluster := parseSystemStatus(status)
+	events := t.Cluster.diffCluster(cluster)
+	for _, event := range events {
+		t.addEvent(event)
+	}
+	t.Cluster = cluster
+}
+
+// GetTimeline returns the current timeline.
+func (t *Timeline) GetTimeline() []string {
+	return t.Timeline
+}
+
+// addEvent appends the provided event to the timeline.
+func (t *Timeline) addEvent(event Event) {
+	if len(t.Timeline) > t.Size {
+		t.Timeline = t.Timeline[1:]
+	}
+	t.Timeline = append(t.Timeline, event.String())
+}
+
+// Cluster represents the overall status of a cluster.
+type Cluster struct {
+	// Status specifies the cluster status.
+	Status string
+	// Nodes specify the individual node statuses.
+	Nodes map[string]*Node
+}
+
+// difCluster calculates the differences from the provided cluster and returns
+// the differences as a list of events.
+func (c *Cluster) diffCluster(cluster *Cluster) []Event {
+	events := []Event{}
+
+	// Compare cluster status
+	if c.Status != cluster.Status {
+		event := NewEvent()
+		event.Description = fmt.Sprintf("cluster status changed from [%s] to [%s]", c.Status, cluster.Status)
```

Comment:
If I was doing this, I would try to decouple the representation of the objects (i.e. descriptions) from the objects themselves. For example, we could have an event object:

```go
type TimelineEvent struct {
  Type string // defines event type, e.g. "status", "nodeAdded", "nodeRemoved", "probeFailed", etc.
  Metadata map[string]string // event-specific metadata e.g. node address, old/new status, etc.
}
```

Then these events can be passed on the wire over gRPC API and the callers can format them appropriately based on the event type and its metadata.